### PR TITLE
Train troops

### PR DESCRIPTION
**Pull request by MoHoseinHabibi**
hi
I added train troop because of my needs and this is how it works:
it calculates max training troop rate(capacity) by count and level of barrack buildings (if they are upgrading won't count them)
it trains troops with max training troop rate(capacity) after it is finished it will claim troops and sleep for 3h (3h because if barrack upgrade possible do that in while time)
if there is not enough resource it will train as much as it's possible
I updated _is_building_upgradeable so if barracks are training troops do not upgrade them

* important note: this does not support tier 5 troops as I don't have their resource requirement

### CODE REVIEW

File: lokbot/farmer.py [hunk 1]
```diff
@@ -821,6 +827,73 @@ def academy_farmer_thread(self, to_max_level=False):
...
 
+    def _troop_training_capacity(self):
+        """
+        return total troop training capacity of all barracks
+        """
+        buildings = self.kingdom_enter.get('kingdom', {}).get('buildings', [])
+        troop_training_capacity = 0
+        for building in buildings:
+            if building.get('state') == BUILDING_STATE_NORMAL:
+                continue
```
Comment (hldh214) [inline lokbot/farmer.py:838]: I think we should use `!=` here?

File: lokbot/farmer.py [hunk 2]
```diff
@@ -821,6 +827,73 @@ def academy_farmer_thread(self, to_max_level=False):
...
+    def train_troop_thread(self, troop_code):
+        current_tasks = self.api.kingdom_task_all().get('kingdomTasks', [])
+
+        worker_used = [t for t in current_tasks if t.get('code') == TASK_CODE_CAMP]
+
+        troop_training_capacity = self._troop_training_capacity()
+
+        if worker_used:
+            if worker_used[0].get('status') == STATUS_CLAIMED:
+                self.api.kingdom_task_claim(random.choice([105, 110]))
```
Comment (hldh214) [inline lokbot/farmer.py:867]: Maybe we need to random choice the position code based on buildings which typed `barrack`, instead of hard-coded 105-110 ?

**Comment by mohammadof313** [comment]
I fixed the if statement and added a random choice position code based on existing buildings

File: lokbot/farmer.py [hunk 3]
```diff
@@ -821,6 +827,83 @@ def academy_farmer_thread(self, to_max_level=False):
...
+                amount = resource // req_resource
+
+        return amount
+
+    def _random_choice_building_position(self, building_code):
+        """
+        return position of a building type by random
+        """
+        buildings = self.kingdom_enter.get('kingdom', {}).get('buildings', [])
+        return random.choice([building['code'] for building in buildings if building['code'] == building_code])
```
Comment (hldh214) [inline lokbot/farmer.py:866]: I think we need a random choice between building's **position** but not **code**, or simply return raw `building` object, and make caller to decide which property to use.

**Review by hldh214 — CHANGES_REQUESTED** [review]

File: lokbot/farmer.py (hunk 3 above)
Comment (mohammadof313) [reply lokbot/farmer.py]: You are right, I was not very focused

**Comment by aliheroo7** [comment]
pls teach us to use this i read but i cant use this bot

**Pull request merged**